Framework: Preload moment locale data when selected language different than `en`

amanya opened the pull request:
This pull request seeks to retake work done by @stephanethomas and @gziolo in https://github.com/Automattic/wp-calypso/pull/12491

## Problem

[`moment`](https://github.com/moment/moment) module is used for localizing dates and times. It comes with lots of `.js` files, one for each supported language. All language files are bundled by Webpack into the `vendor` chunk.

![image](https://user-images.githubusercontent.com/4490611/31306553-d5a616b8-ab52-11e7-86ca-5e0d7ac0662e.png)

## Solving

* Setup Wepbpack to ignore bundling of `moment` locales, as explained in [`IgnorePlugin`'s documentation](https://webpack.js.org/plugins/ignore-plugin/#ignore-moment-locales).

* Change the server code to load the correct locale file before the chunk is loaded, as explained [here](https://github.com/Automattic/wp-calypso/pull/12491#issuecomment-303736837).

* Setup Webpack to use a custom loader to patch `i18n-calypso` module. `i18n-calyspo` code is changed so that if locale is different than `en`, use `require` with a [`bundle-loader`](https://www.npmjs.com/package/bundle-loader) query to request the locale bundle. That will use a callback to set the correct locale when the locale chunk is loaded.

## Testing

* Run git checkout try/lazyload-moment and start your server, or open a live branch
* Log in to an English account
* Check that the vendor bundle is now ~2.3M (instead of ~2.5M) in the network panel of your browser
* Check that no moment-locale bundle was loaded in this panel
* Navigate to the Account Settings page
* Notice the date in English just below the Username field (e.g. Joined April 2016)
* Select a different language in the Interface Language select box
* Hit the Save Account Settings button, which will reload the page
* Check that a moment-locale was loaded this time in the network panel
* Check that the date below the Username field is now in the language chosen

Note that in logged out routes, the language will always be English due to [this](https://github.com/Automattic/wp-calypso/issues/14508) bug.

stephanethomas commented on package.json:
```diff
@@ -38,6 +38,7 @@
     "body-parser": "1.17.2",
     "bounding-client-rect": "1.0.5",
     "browser-filesaver": "1.1.0",
+    "bundle-loader": "0.5.5",
```
We should [regenerate `npm-shrinkwrap.json`](https://github.com/Automattic/wp-calypso/blob/master/docs/shrinkwrap.md) since `package.json` has been updated.

amanya commented:
@stephanethomas I added your suggestions and rebased against master.

About the bug you mention, I see that it also happens in master branch (I'm seeing it in production wordpress.com also), probably related to this warning when loading the page:

```
warning.js:33 Warning: Failed prop type: The prop `locale` is marked as required in `Login`, but its value is `undefined`.
    in Login (created by Localized(Login))
    in Localized(Login) (created by Connect(Localized(Login)))
    in Connect(Localized(Login))
    in div (created by LayoutLoggedOut)
    in div (created by LayoutLoggedOut)
    in div (created by LayoutLoggedOut)
    in LayoutLoggedOut (created by Connect(LayoutLoggedOut))
    in Connect(LayoutLoggedOut) (created by ReduxWrappedLayout)
    in Provider (created by ReduxWrappedLayout)
    in ReduxWrappedLayout
```

I'll open an issue to track it separately if you think its a good idea

stephanethomas commented:
> About the bug you mention, I see that it also happens in master branch (I'm seeing it in production wordpress.com also),

I missed the last part of your comment, and created https://github.com/Automattic/wp-calypso/issues/18866 to address that issue.

nb commented on server/pages/index.jade:
```diff
@@ -1,12 +1,12 @@
 	<3
-	             _
-	    ___ __ _| |_   _ _ __  ___  ___
+				 _
```
What changed here?

amanya replied:
FIxed! I converted spaces in the logo to tabs by mistake

yurynix commented on server/render/index.js:
```diff
@@ -89,6 +90,10 @@ export function serverRender( req, res ) {
 		context.i18nLocaleScript = '//widgets.wp.com/languages/calypso/' + context.lang + '.js';
 	}
 
+	if ( moment.locale() !== 'en' ) {
```
For my own understanding, if you could clarify those, it'll be great:

1. Where do we change the output of `moment.locale()` ?
2. Can it be anything other than english on the server?

Thank you!

amanya replied:
1.`moment.locale()` is initialized by `i18n-calypso` by means of the router setup: https://github.com/Automattic/wp-calypso/blob/master/client/login/index.web.js#L27 when calling `/log-in` with a locale different than `en`

2. Yes, is set by `i18n-calypso`: https://github.com/Automattic/i18n-calypso/blob/master/lib/index.js#L193

amanya replied:
@yurynix I applied your suggestions: removed `moment` import and used `context.lang` to initialize moment locale file name.

It seems to work correctly :+1:

yurynix commented on server/bundler/i18n-calypso-loader.js:
```diff
@@ -0,0 +1,19 @@
+	this.cacheable && this.cacheable();
+
+	if ( this.resource.endsWith( 'node_modules/i18n-calypso/lib/index.js' ) ) {
+		const newContent = content.replace(
```
Just a question, I'm probably missing something obvious, why do we need to do that via loader and not put that directly into [i18n-calypso](https://github.com/Automattic/i18n-calypso) ?

amanya replied:
Because we generate Webpack code that will be translated again by Webpack to:
```	
    // Block injected at build time via i18n-calypso-loader Webpack loader
	if ( this.state.localeSlug === 'en' ) {
		// Moment ships with the English locale by default, no need to load anything in that case
	} else {
		__webpack_require__("./node_modules/moment/locale recursive ./node_modules/bundle-loader/index.js?name=moment-locale-[name]!./ ^\\.\\/.*$")("./" + this.state.localeSlug)( () => {
			moment.locale( this.state.localeSlug );
		} );
	}
```

yurynix replied:
Aha, right, thank you for explaining!

stephanethomas replied:
Also we may want to use `i18n-calypso` without Webpack ;).

yurynix commented:
Thank you for your work on this @amanya , I've left a few questions =)

stephanethomas commented:
https://github.com/Automattic/wp-calypso/pull/18924 has been merged, let's rebase this pull request and do some testing :).

amanya commented:
@stephanethomas done :+1: I went through the test workflow and seems to work fine

amanya commented:
I've tried replacing the `require( bundle-loader...` stuff by

```
	    import(
			/* webpackChunkName: moment/locale/[request] */
			'moment/locale/' + this.state.localeSlug
		).then( _ => {
			moment.locale( this.state.localeSlug );
		});
```

and although it seems to do the job, I cannot manage to name the chunks properly. All of them are named `ca.js.js` and using the `require( bundle-loader` statement were named `moment-locale-ca.js`.

I suspect that is caused because then I'm not using `bundle-loader` and I don't know if I can do the same using Webpack's import.

Some context: https://github.com/webpack/webpack/issues/4807#issuecomment-299158833

samouri commented on server/bundler/i18n-calypso-loader.js:
```diff
@@ -0,0 +1,24 @@
+ */
+
+module.exports = function( content ) {
+	this.cacheable && this.cacheable();
```
this line is no longer necessary in webpack2+ since cacheable is [enabled by default](https://webpack.js.org/api/loaders/#this-cacheable)

stephanethomas replied:
This comment has not been addressed @amanya.

amanya replied:
I forgot about it, it is removed now

amanya commented on server/bundler/i18n-calypso-loader.js:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Webpack loader used to inject Webpack code into `i18n-calypso` module to
+ * load asyncronously the current locale.
```
> Looking at the documentation maybe it's because of double quotes that are missing?

You are right, good catch! Now the chunk name is generated correctly :) I've committed the changes.

amanya replied:
> [...] Are the locale bundles correctly minified in that case?

I pushed a new commit with changes in the minifier that makes it traverse all `./public` subdirectories to ensure that locale files are minified.

I tested it and now it works correctly.

amanya commented:
Minified vendor bundle went from 932Kb to 815Kb :)

samouri commented on bin/minify.js:
```diff
@@ -12,23 +12,31 @@ const os = require( 'os' );
-const files = fs
-	.readdirSync( './public' )
+
+const walkDir = ( dir ) => fs
```
whats the reason for this change?

amanya replied:
Since we don't bundle the moment-locale files into the vendor chunk, we need to load them from the `public/moment/locale` directory. As the minifier only processed files in the `public/` directory, I did this changes, to also minify files in all child subdirectories.

samouri replied:
Thank you for the helpful explanation! Totally makes sense ⭐️

samouri commented on server/bundler/i18n-calypso-loader.js:
```diff
@@ -0,0 +1,27 @@
+ */
+
+module.exports = function( content ) {
+	this.cacheable && this.cacheable();
```
this line isn't necessary. `cacheable: true` is the default in webpack2+

samouri commented:
I tested this out in both docker and local, both checked out. was pretty cool to see locale data loaded from `http://wpcalypso.wordpress.com/calypso/moment/locale/es.c2ee057859d8bcda0c99.min.js`.

I have on architecture question:

The current design of the solution has a custom webpack loaded that makes a runtime async-chunk-load for locale data. Would it be both simpler and more performance to ditch the runtime loader in favor of an additional script tag in the `index.jade` file?

amanya commented on server/pages/index.jade:
```diff
@@ -236,6 +236,8 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 		if chunk
 			script(src=urls[ chunk ])
+		if momentLocaleFile
+			script(src=urls[ momentLocaleFile ])
```
@stephanethomas I'm not sure that this code is needed, at least it works correctly without it in the test flow I specified in the PR description.

I think that the original purpose of that is to have the moment-locale data loaded when changing languages as discussed [here](https://github.com/Automattic/wp-calypso/pull/12491#issuecomment-304229400) but I don't entirely see the point because when you change the language the page is reloaded, so `i18n-moment` takes care of the loading again.

Wdyt?

amanya replied:
I see, thanks for the explanation!

Tug commented:
So, I ran some benchmarking locally using chrome devtools on a production env.

In terms of time to full page render, I did not see any significant difference between this branch and master, with or without caching. This is mainly because loading the rest proxy iframe dwarfs all other assets loading time (1.35s on average for me). The difference I saw between the 2 (around 300ms on average) were actually in favour of the master version. 

The good news is that our new chunk `/calypso/moment/locale/{locale_slug}.json` is able to load in parallel of that iframe. However the difference in loading time we gain with a lighter version of the vendor bundle (2.6MB instead of 2.7MB) is negligible compared to that extra request the browser has to handle (even though it is not very costly because it runs in parallel with other assets).

Some screenshots:
- Loading this branch without cache
![image](https://user-images.githubusercontent.com/230230/32741446-904ecf12-c8a6-11e7-861b-87bd061ec36a.png)

- Loading master without cache
![image](https://user-images.githubusercontent.com/230230/32741472-a219bdce-c8a6-11e7-82ca-e6a4090d64fd.png)

In the end I'm not sure this improvement is actually worth it as it increase the complexity of the build and delays the loading time of the locale for moment.js. Something we haven't considered but making our call to `moment.locale()` asynchronous may cause problems if some function use moment-timezone on load.

amanya commented:
@Tug Thanks for the review. I agree that looking at the numbers, the changes proposed in that PR are not worth the added complexity and likelihood of new errors. I'm closing the PR
